sharding/node/backend.go

```diff
@@ -0,0 +1,198 @@
+// Ethereum network.
+type ShardEthereum struct {
+	shardConfig  *params.ShardConfig  // Holds necessary information to configure shards.
+	txPool       *txpool.ShardTXPool  // Defines the sharding-specific txpool. To be designed.
+	actor        sharding.Actor       // Either notary, proposer, or observer.
+	shardChainDb ethdb.Database       // Access to the persistent db to store shard data.
+	eventFeed    *event.Feed          // Used to enable P2P related interactions via different sharding actors.
+	smcClient    *mainchain.SMCClient // Provides bindings to the SMC deployed on the Ethereum mainchain.
+
+	// Lifecycle and service stores.
+	serviceFuncs []sharding.ServiceConstructor     // Service constructors (in dependency order).
+	services     map[reflect.Type]sharding.Service // Currently running services.
+	lock         sync.RWMutex
+}
+
+// New creates a new sharding-enabled Ethereum instance. This is called in the main
+// geth sharding entrypoint.
+func New(ctx *cli.Context) (*ShardEthereum, error) {
+
+	shardEthereum := &ShardEthereum{}
+
+	path := node.DefaultDataDir()
+	if ctx.GlobalIsSet(utils.DataDirFlag.Name) {
+		path = ctx.GlobalString(utils.DataDirFlag.Name)
+	}
+
+	endpoint := ctx.Args().First()
+	if endpoint == "" {
+		endpoint = fmt.Sprintf("%s/%s.ipc", path, mainchain.ClientIdentifier)
+	}
+	if ctx.GlobalIsSet(utils.IPCPathFlag.Name) {
+		endpoint = ctx.GlobalString(utils.IPCPathFlag.Name)
+	}
+
+	passwordFile := ctx.GlobalString(utils.PasswordFileFlag.Name)
+	depositFlag := ctx.GlobalBool(utils.DepositFlag.Name)
+	actorFlag := ctx.GlobalString(utils.ActorFlag.Name)
+
+	smcClient, err := mainchain.NewSMCClient(endpoint, path, depositFlag, passwordFile)
+	if err != nil {
+		return nil, err
+	}
+
+	// Adds the initialized SMCClient to the ShardEthereum instance.
+	shardEthereum.smcClient = smcClient
+
+	// TODO: Find a better way to do this???
+	if err := shardEthereum.registerP2P(); err != nil {
+		return nil, err
+	}
+
+	if err := shardEthereum.registerTXPool(actorFlag); err != nil {
+		return nil, err
+	}
+
+	if err := shardEthereum.registerActorService(actorFlag); err != nil {
+		return nil, err
+	}
+
+	return shardEthereum, nil
+}
+
+// Start the ShardEthereum service and kicks off the p2p and actor's main loop.
+func (s *ShardEthereum) Start() error {
+	log.Info("Starting sharding node")
+
+	services := make(map[reflect.Type]sharding.Service)
+	for _, constructor := range s.serviceFuncs {
+		// Create a new context for the particular service.
+		ctx := &sharding.ServiceContext{
+			Services: make(map[reflect.Type]sharding.Service),
+		}
+		// Copy needed for threaded access.
+		for kind, s := range services {
+			ctx.Services[kind] = s
+		}
+		// Construct and save the service.
+		service, err := constructor(ctx)
+		if err != nil {
+			return err
+		}
+		kind := reflect.TypeOf(service)
+		if _, exists := services[kind]; exists {
+			return fmt.Errorf("service already exists: %v", kind)
+		}
+		services[kind] = service
```

Comment:
Quick question: Why do we build the service contexts and mapping here? 
Why isn't it done as part of `Register`?

There might be a good reason. If so, could you add a quick comment to the start of this logic?